tests/Crypto/DecryptAllTest.php

```diff
@@ -182,17 +186,184 @@ public function testPrepareLoginException($loginType) {
+			->method('getMasterKeyPassword')
+			->willReturn('pass');
+		$this->keyManager->expects($this->once())
+			->method('getRecoveryKeyId')
+			->willReturn('user1');
+		if (!$expectedResult) {
+			$this->crypt->expects($this->once())
+				->method('decryptPrivateKey')
+				->willReturn(false);
+		}
+
+		$result = $this->instance->prepare($input, $output, 'user1');
+		$this->assertEquals($expectedResult, $result);
+	}
+
+	public function providesUserKeyPrepareResult() {
+		return [
+			[true, 'user1'],
+			[false, 'user1'],
+			[true, ''],
+			[false, '']
+		];
+	}
+
+	/**
+	 * @dataProvider providesUserKeyPrepareResult
+	 */
+	public function testPrepareUserKeyWithRecoveryPasswd($expectedResult, $user) {
+		$input = $this->createMock(InputInterface::class);
+		$output = $this->createMock(OutputInterface::class);
+
+		$this->util->expects($this->once())
+			->method('isMasterKeyEnabled')
+			->willReturn(false);
+
+		if (!(($expectedResult === false) && ($user === ''))) {
+			$input->expects($this->exactly(2))
+				->method('getOption')
+				->with('method')
+				->willReturn('recovery');
+			\putenv("OC_RECOVERY_PASSWORD=foo");
 		} else {
+			$input->expects($this->exactly(1))
+				->method('getOption')
+				->with('method')
+				->willReturn('foo');
+		}
+
+		if (!$expectedResult && ($user !== '')) {
+			$this->util->expects($this->once())
+				->method('isRecoveryEnabledForUser')
+				->willReturn(true);
+
 			$this->keyManager->expects($this->once())
 				->method('checkRecoveryPassword')
-				->willThrowException(new \Exception());
-			$this->instance->prepare($input, $output, 'xyz');
+				->willReturn(true);
+
+			$this->keyManager->expects($this->exactly(2))
+				->method('getRecoveryKeyId')
+				->willReturn($user);
+		}
+
+		if (($expectedResult === true) && ($user === '')) {
+			$this->keyManager->expects($this->exactly(1))
+				->method('getRecoveryKeyId')
+				->willReturn($user);
 		}
 
+		if ($expectedResult === true) {
+			$this->crypt->expects($this->once())
+				->method('decryptPrivateKey')
+				->willReturn(true);
+		} else {
+			if (!(($expectedResult === false) && ($user === ''))) {
+				$this->crypt->expects($this->once())
+					->method('decryptPrivateKey')
+					->willReturn(false);
+			}
+		}
+
+
+		$result = $this->instance->prepare($input, $output, $user);
+		$this->assertEquals($expectedResult, $result);
+	}
+
+	/**
+	 * @dataProvider providesMasterKeyPrepareResult
+	 */
+	public function testPrepareUserKeyWithOCPasswd($expectedResult) {
+		$input = $this->createMock(InputInterface::class);
+		$output = $this->createMock(OutputInterface::class);
+
+		$this->util->expects($this->once())
+			->method('isMasterKeyEnabled')
+			->willReturn(false);
+
+		$input->expects($this->once())
+			->method('hasOption')
+			->willReturn(true);
+		$input->expects($this->exactly(2))
+			->method('getOption')
+			->with('method')
+			->willReturnOnConsecutiveCalls(null, 'password');
+		\putenv("OC_PASSWORD=foo");
```

Comment:
how about extending core's "EnvironmentHelper" and adding a method for reasing env variables ?

this means you'd be able to mock it in the tests

Reply:
Added a method in EnvironmentHelper. And in the tests its being mocked.